### PR TITLE
Pool locking mechanism

### PR DESCRIPTION
**Pull request by lukasz-zimnoch**
Sortition pool should expose a way to enable and disable the locked state in which operators cannot be inserted, removed, and updated. The first use case is beacon v2 where `RandomBeacon` contract needs to lock the pool during group creation process.

### CODE REVIEW

File: contracts/SortitionPool.sol [hunk 1]
```diff
@@ -35,15 +42,33 @@ contract SortitionPool is SortitionTree {
...
+      false
     );
   }
 
+  /// @notice Locks the sortition pool. In locked state, members cannot be
+  ///         inserted and removed from the pool. Members statuses cannot
+  ///         be updated as well.
+  /// @dev Can be called only by the contract owner.
+  function lock() public {
+    require(msg.sender == poolParams.owner, "Caller is not the owner");
```
Comment (pdyraga) [inline contracts/SortitionPool.sol:55]: Can we take the opportunity and extract it to `onlyOwner` modifier?

File: contracts/SortitionPool.sol [hunk 2]
```diff
@@ -21,10 +21,17 @@ contract SortitionPool is SortitionTree {
     uint256 minimumStake;
     uint256 poolWeightDivisor;
     address owner;
+    bool locked;
```
Comment (pdyraga) [inline contracts/SortitionPool.sol:24]: What if we unwrap all these parameters from `PoolParams` structure? I think in the past we were passing this struct as a parameter to some functions, now it is no longer the case. Additionally, `stakingContract` and `poolWeightDivisor` could be made `immutable` to save on gas.

Comment (pdyraga) [reply contracts/SortitionPool.sol:24]: Can we also make `minimumStake` governable by owner?

File: test/sortitionPoolTest.js [hunk 3]
```diff
@@ -39,32 +39,93 @@ describe("SortitionPool", () => {
...
+      // Pool is unlocked by default so we need to lock it explicitly.
+      await pool.connect(owner).lock()
+    })
+
+    context("when called by the owner", () => {
+      beforeEach(async () => {
+        await pool.connect(owner).unlock()
+      })
+
+      it("should lock the pool", async () => {
```
Comment (pdyraga) [inline test/sortitionPoolTest.js:74]: *unlock

File: contracts/SortitionPool.sol (hunk 1 above)
Comment (lukasz-zimnoch) [reply contracts/SortitionPool.sol]: Done.

File: test/sortitionPoolTest.js (hunk 3 above)
Comment (lukasz-zimnoch) [reply test/sortitionPoolTest.js]: Done.

File: contracts/SortitionPool.sol [hunk 4]
```diff
@@ -17,31 +17,72 @@ contract SortitionPool is SortitionTree, Ownable {
   using DynamicArray for DynamicArray.UintArray;
   using DynamicArray for DynamicArray.AddressArray;
 
-  struct PoolParams {
-    IStaking stakingContract;
-    uint256 minimumStake;
-    uint256 poolWeightDivisor;
+  /// @notice Governance delay that needs to pass before any risk manager
+  ///         parameter change initiated by the governance takes effect.
+  uint256 public constant GOVERNANCE_DELAY = 48 hours;
```
Comment (pdyraga) [inline contracts/SortitionPool.sol:22]: Let's do it without governable delay. If we need a governable delay, we will add it in `RandomBeacon` or application controlling the pool.

Comment (lukasz-zimnoch) [reply contracts/SortitionPool.sol]: Ok, done: https://github.com/keep-network/sortition-pools/pull/132/commits/de170130305d56bc9218fb811a46a9e09e1e646d

File: test/sortitionPoolTest.js [hunk 5]
```diff
@@ -117,22 +192,119 @@ describe("SortitionPool", () => {
...
+        it("should update operators status", async () => {
+          const group = await pool.connect(owner).selectGroup(5, seed)
+          await pool.connect(owner).nonViewSelectGroup(5, seed)
+          expect(group).to.deep.equal([
+            bob.address,
+            bob.address,
+            bob.address,
+            bob.address,
+            bob.address,
+          ])
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply test/sortitionPoolTest.js]: I just moved that case to get a proper test structure and didn't dive too deep here. Changed the assertion: https://github.com/keep-network/sortition-pools/pull/132/commits/e0324093c320d3b6d4f7d0f1a44bff4c178d0b74

File: test/sortitionPoolTest.js [hunk 6]
```diff
@@ -229,18 +229,13 @@ describe("SortitionPool", () => {
...
       await pool.connect(owner).insertOperator(alice.address)
-      await pool.connect(owner).insertOperator(bob.address)
 
       aliceID = await pool.getOperatorID(alice.address)
-      bobID = await pool.getOperatorID(bob.address)
 
-      await staking.setStake(bob.address, 390000)
       await staking.setStake(alice.address, 1000)
 
       await helpers.time.mineBlocks(11)
```
Comment (pdyraga) [inline test/sortitionPoolTest.js:241]: Is this needed?

Comment (lukasz-zimnoch) [reply test/sortitionPoolTest.js]: Removed https://github.com/keep-network/sortition-pools/pull/132/commits/f1524a44fe78726894ece27a993ec80aba5b4fbb

**Review by pdyraga — APPROVED** [review]

**Pull request merged**